### ISSUE TITLE
0.58.2 crashes on grants

### ISSUE THREAD

**Issue by bennylu2** [Snowflake-Labs/terraform-provider-snowflake#1627]
**Provider Version**

`0.58.2`

**Terraform Version**

`1.3.4`

**Describe the bug**

Provider is crashing on grant resources with an index out of range error

```
...

│ Error: Request cancelled
│ 
│   with module.tdm_demo.module.private_bdi_internal_monitor.module.stage_this["USAGE"].snowflake_stage_grant.this,
│   on .terraform/modules/tdm_demo/tdm/modules/stage-grant/main.tf line 1, in resource "snowflake_stage_grant" "this":
│    1: resource "snowflake_stage_grant" "this" {
│ 
│ The plugin.(*GRPCProvider).UpgradeResourceState request was cancelled.
╵

Stack trace from the terraform-provider-snowflake_v0.58.2 plugin:

panic: runtime error: index out of range [5] with length 5

goroutine 979 [running]:
github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/resources.ParseDatabaseGrantID({0xc0005b6450, 0x16})
        github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/resources/database_grant.go:241 +0x3ca
github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/resources.ReadDatabaseGrant(0xc000684b00, {0x130ce60, 0xc00044d2b0})
        github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/resources/database_grant.go:126 +0xef
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).read(0x1580ae0?, {0x1580ae0?, 0xc000e0ba70?}, 0xd?, {0x130ce60?, 0xc00044d2b0?})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.25.0/helper/schema/resource.go:712 +0x178
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).RefreshWithoutUpgrade(0xc0003f88c0, {0x1580ae0, 0xc000e0ba70}, 0xc000c1d2b0, {0x130ce60, 0xc00044d2b0})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.25.0/helper/schema/resource.go:1015 +0x585
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ReadResource(0xc0003c9e78, {0x1580ae0?, 0xc000e0b950?}, 0xc000801140)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.25.0/helper/schema/grpc_provider.go:613 +0x4a5
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ReadResource(0xc00039a960, {0x1580ae0?, 0xc000e0b1a0?}, 0xc0001c9b00)
        github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov5/tf5server/server.go:748 +0x4b1
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ReadResource_Handler({0x12d8d00?, 0xc00039a960}, {0x1580ae0, 0xc000e0b1a0}, 0xc0001cbab0, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:349 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0003f2000, {0x15868a0, 0xc000510d00}, 0xc000c98120, 0xc000615230, 0x1d36570, 0x0)
        google.golang.org/grpc@v1.53.0/server.go:1336 +0xd23
google.golang.org/grpc.(*Server).handleStream(0xc0003f2000, {0x15868a0, 0xc000510d00}, 0xc000c98120, 0x0)
        google.golang.org/grpc@v1.53.0/server.go:1704 +0xa2f
google.golang.org/grpc.(*Server).serveStreams.func1.2()
        google.golang.org/grpc@v1.53.0/server.go:965 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
        google.golang.org/grpc@v1.53.0/server.go:963 +0x28a

Error: The terraform-provider-snowflake_v0.58.2 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.

Operation failed: failed running terraform plan (exit 1)
```

**Expected behavior**

No crash

**Comment by ogr003** [comment]
This bug was introduced in 0.56.4 and has persisted since. Providing a proper fix for this should be given top priority as there is no use in introducing new features if this bug stops us from being able to update anyway.

**Comment by sfc-gh-swinkler** [comment]
What is the ID of your stage grant resource? And are there any other grant resources causing issue?

**Comment by bennylu2** [comment]
ID for stage

```
tf state show 'module.tdm_demo.module.private_bdi_internal_monitor.module.stage_this["USAGE"].snowflake_stage_grant.this'
# module.tdm_demo.module.private_bdi_internal_monitor.module.stage_this["USAGE"].snowflake_stage_grant.this:
resource "snowflake_stage_grant" "this" {
    database_name          = "TDM_DEMO"
    enable_multiple_grants = true
    id                     = "TDM_DEMO|BDI_INTERNAL_MONITOR||USAGE||false"
    on_future              = true
    privilege              = "USAGE"
    roles                  = [
        "TDM_DEMO__BDI_INTERNAL_MONITOR__READ",
        "TDM_DEMO__BDI_INTERNAL_MONITOR__WRITE",
    ]
    schema_name            = "BDI_INTERNAL_MONITOR"
    with_grant_option      = false
}
```

**Comment by bennylu2** [comment]
> panic: runtime error: index out of range [5] with length 5

Likely this https://github.com/Snowflake-Labs/terraform-provider-snowflake/blob/711ac0cbc886bf8be6a5a2651234df778452b9df/pkg/resources/database_grant.go#L248

**Comment by sfc-gh-swinkler** [comment]
@bennylu2 i apologize for the trouble this has caused you. Theres multiple versions of IDs, and this is one of the older versions which I had not anticipated. I have a patch to fix it in 0.59, and hopefully this should be the last of the crashes you experience with regards to this. Please open a new PR if you encounter any further problems. I have a more detailed discussion about this change here, explaining the motivations and challenges in fixing the problems: https://github.com/Snowflake-Labs/terraform-provider-snowflake/discussions/1633.

**Comment by bennylu2** [comment]
@sfc-gh-swinkler thanks for staying on top of this. Most of the grant resources now show no crash or error. However, we are getting occasional errors on the `snowflake_role_grants` resource, opened #1642
